geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/LiveFeatureCacheCQEngine.scala

```diff
@@ -0,0 +1,117 @@
+import org.opengis.filter._
+
+import scala.collection.JavaConversions._
+import scala.collection.JavaConverters._
+import scala.collection.mutable
+
+class LiveFeatureCacheCQEngine(sft: SimpleFeatureType,
+                               expirationPeriod: Option[Long])(implicit ticker: Ticker)
+  extends LiveFeatureCache with LazyLogging {
+
+  val geocq = new GeoCQEngine(sft)
+
+  // JNH: Do we need FeatureHolder anymore?
+  private val cache: Cache[String, FeatureHolder] = {
+    val cb = CacheBuilder.newBuilder().ticker(ticker)
+    expirationPeriod.foreach { ep =>
+      cb.expireAfterWrite(ep, TimeUnit.MILLISECONDS)
+        .removalListener(new RemovalListener[String, FeatureHolder] {
+          def onRemoval(removal: RemovalNotification[String, FeatureHolder]) = {
+            if (removal.getCause == RemovalCause.EXPIRED) {
+              println(s"Removing feature ${removal.getKey} due to expiration after ${ep}ms")
+
+              logger.debug(s"Removing feature ${removal.getKey} due to expiration after ${ep}ms")
+              val ret = geocq.remove(removal.getValue.sf)
+              println(s"Removing feature ${removal.getKey} due to expiration after ${ep}ms returned $ret from CQEngine")
+            }
+          }
+        })
+    }
+    cb.build()
+  }
+
+  val features: mutable.Map[String, FeatureHolder] = cache.asMap().asScala
+
+  def size(): Int = {
+    features.size
+  }
+
+  def size(f: Filter): Int = {
+    if (f == Filter.INCLUDE) {
+      features.size
+    } else {
+      getReaderForFilter(f).getIterator.length
+    }
+  }
+
+  override def cleanUp(): Unit = { cache.cleanUp() }
+
+  override def createOrUpdateFeature(update: CreateOrUpdate): Unit = {
+    val sf = update.feature
+    val id = sf.getID
+    val old = cache.getIfPresent(id)
+    if (old != null) {
+      geocq.remove(old.sf)
+    }
+    val env = sf.geometry.getEnvelopeInternal
+    geocq.add(sf)
+    cache.put(id, FeatureHolder(sf, env))
+  }
+
+  override def getFeatureById(id: String): FeatureHolder = cache.getIfPresent(id)
+
+  override def removeFeature(toDelete: Delete): Unit = {
+    val id = toDelete.id
+    val old = cache.getIfPresent(id)
+    if (old != null) {
+      //spatialIndex.remove(old.env, old.sf)
+      geocq.remove(old.sf)
+      cache.invalidate(id)
+    }
+  }
+
+  override def clear(): Unit = {
+    cache.invalidateAll()
+    geocq.clear()        // Consider re-instanting the CQCache
+  }
+
+  def getReaderForFilter(filter: Filter): FR =
+    filter match {
+      case f: Id            => fid(f)
+      case f                => geocq.getReaderForFilter(f)
+      // JNH: Consider testing filter rewrite before passing to CQEngine?
+    }
+
+  def fid(ids: Id): FR = {
+    println("Queried for IDs; using Guava ID index")
+    val iter = ids.getIDs.flatMap(id => features.get(id.toString).map(_.sf)).iterator
+    new DFR(sft, new DFI(iter))
+  }
+
+  // TODO: Remove after testing is finished
```

Comment:
My thought was that passing a CQEngine query to the story was unneeded.  Any opinions from others?